### ISSUE TITLE
Full support for partial imports

### ISSUE THREAD

**Issue by donat** [eclipse/buildship#247]
<a href="https://github.com/oehme"><img src="https://avatars.githubusercontent.com/u/1191797?v=3" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [oehme](https://github.com/oehme)**
_Saturday Apr 30, 2016 at 13:20 GMT_

----

## Motivation

Users often have very large builds and only want to work on a subset of that build. Nevertheless, that subset might have dependencies on other projects. If those projects are not imported, the classpath of the dependents is incomplete. This means that users have to import much more projects than they are interested in changing.

We should substitute those project dependencies for external dependencies on the locally built jar of those projects. The jar task should be automatically invoked as part of the project synchronization. 
## Proposed Change
- The TAPI needs to provide information how to build a substitute for a project dependency if that project is not imported into the workspace
- Buildship replaces the project dependency in the classpath with an external dependency on the locally built jar
- The jar is built as part of the project synchronization
## Code review
- [ ] Does the change work?
  - [ ] Try out the feature, executing it like a end-user would.
  - [ ] Check corner cases and try to break it in various ways.
  - [ ] Are the error messages informative?
- [ ] Is the change sufficiently tested?
  - [ ] Are there appropriate levels of unit & integration tests?
  - [ ] Do the tests indicate what they are actually testing?
  - [ ] Are the tests in the right location?
  - [ ] Can new tests be consolidated with any existing tests?
- [ ] Is it sufficiently documented?
  - [ ] Is the user guide documentation clear and informative?
  - [ ] Do the release notes clearly convey the value of the feature?
- [ ] Is it implemented well?
  - [ ] Does it minimize surface area (particularly public)?
  - [ ] Does it follow our coding conventions?
  - [ ] Does it use appropriate data structures?
  - [ ] Does it minimize potential for invalid states?
  - [ ] Does it suitably reuse existing infrastructure?
  - [ ] Does it suitably consider concurrency?
  - [ ] Does it suitably consider failure modes?
  - [ ] Does it suitably consider performance?
  - [ ] Does it introduce new types or move existing ones? Are they located in the appropriate package? (e.g. internal classes in internal packages)

**Comment by donat** [comment]
https://bugs.eclipse.org/bugs/show_bug.cgi?id=487159

**Comment by donat** [comment]
The `EclipseModelBuilder` already takes into account the state of the Eclipse workspace and substitutes the project dependencies with publications in case the corresponding Eclipse workspace project is closed.

We only need to reuse this functionality and provide a UI for selecting filtered projects.

**Comment by rotty3000** [comment]
A UI element for filtering has limitations. Reason being that if you apply a filter during import then every time you refresh that filter also needs to be respected which means data needs to be stored in the eclipse properties system. I think it's much more powerful and easier to have these details be part of the gradle build like is already supported via `eclipse` plugin project configurations, perhaps with a `buildShip` configuration. Anyway, I just see it as being easier and more powerful in a gradle closure, but any solution that is reliable would be fine with me.

**Comment by donat** [comment]
Good point. I'm not against adding a new configuration option to the Gradle dsl, it can go hand-in-hand.

**Comment by loetifuss** [comment]
> Buildship replaces the project dependency in the classpath with an external dependency on the locally built jar
It would be nice if Buildship could also support replacing a project dependency with a published artifact in a maven repository (e.g. latest snapshot or some configurable version). Imagine a complex build with a lot of interdependent projects: building all those external dependencies locally would slow down the build unnecessarily.
